File: reader.py
from __future__ import print_function

import datetime
import argparse
import os.path
import time
import math


from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials as GoogleCredentials
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
from google.auth.exceptions import RefreshError

import asyncio

from viam.robot.client import RobotClient
from viam.rpc.dial import Credentials as ViamCredentials, DialOptions
from viam.components.motor import Motor

# If modifying these scopes, delete the file token.json.
SCOPES = ['https://www.googleapis.com/auth/calendar.readonly']

OUT_OF_OFFICE = 0
WORK_FROM_HOME = 1
GOING_TO_EVENT = 2
FOCUS_TIME = 3
AVAILABLE = 4
IN_MEETING = 5

event_type_to_wheel_position = {
    "outOfOffice": 0,
    "focusTime": 3,
    "default": 5
}

async def connect(location_secret: str, robot_address: str):
    creds = ViamCredentials(
        type='robot-location-secret',
        payload=location_secret,)
    opts = RobotClient.Options(
        refresh_interval=0,
        dial_options=DialOptions(credentials=creds, timeout=10.0),
        # not available in SDK yet (pending release)
        # timeout=5
    )
    for x in range(50):
        try:
            print("connection try", x)
            return await RobotClient.at_address(robot_address, opts)
        except Exception as e:
            print("Failed to connect. Please try hitting RESET on ESP32", e)
            pass
    raise Exception("Too many connection attempts to robot failed. Please sure that robot is on and connected to wifi.")

def get_next_wheel_position() -> int:
    """Shows basic usage of the Google Calendar API.
    Prints the start and name of the next 10 events on the user's calendar.
    """
    creds = None
    # The file token.json stores the user's access and refresh tokens, and is
    # created automatically when the authorization flow completes for the first
    # time.
    if os.path.exists('token.json'):
        creds = GoogleCredentials.from_authorized_user_file('token.json', SCOPES)
    # If there are no (valid) credentials available, let the user log in.
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
                try:
                    creds.refresh(Request())
                except RefreshError:
                    creds = get_creds()
        else:
            creds = get_creds()
        # Save the credentials for the next run
        with open('token.json', 'w') as token:
            token.write(creds.to_json())

    try:
        service = build('calendar', 'v3', credentials=creds)

        # Call the Calendar API
        now = datetime.datetime.utcnow().isoformat() + 'Z'  # 'Z' indicates UTC time
        print('Getting the upcoming 10 events')
        events_result = service.events().list(calendarId='primary', timeMin=now,
                                              maxResults=1, singleEvents=True,
                                              orderBy='startTime', eventTypes=["default", "focusTime", "outOfOffice", "workingLocation"]).execute()
        events = events_result.get('items', [])

        if not events:
            print('No upcoming events found.')
            return

        event = events[0]
        event_type = event['eventType']
        print("next event type:", event_type)
        start = event['start'].get('dateTime', event['start'].get('date'))
        start_date = datetime.datetime.fromisoformat(start)
        if event_type == 'workingLocation':
            if event['summary'] != "Office":
                return WORK_FROM_HOME
        elif start_date >= datetime.datetime.now(start_date.tzinfo) + datetime.timedelta(minutes=5):
            print("next event is > 5min from now, so AVAILABLE")
            return AVAILABLE
        elif start_date > datetime.datetime.now(start_date.tzinfo):
            print("next event is <= 5min from now, so GOING_TO_EVENT")
            return GOING_TO_EVENT
        wheel_position = event_type_to_wheel_position.get(event_type)
        return wheel_position

    except HttpError as error:
        print('An error occurred: %s' % error)

def get_creds():
    flow = InstalledAppFlow.from_client_secrets_file(
        'credentials.json', SCOPES)
    return flow.run_local_server(port=0)

async def control_wheel(wheel_motor: Motor, current_wheel_position: int) -> (int, Exception):
    next_wheel_position = get_next_wheel_position()
    if current_wheel_position != next_wheel_position:
        print("turning wheel from", current_wheel_position, " to position ", next_wheel_position)    
        slices = (current_wheel_position - next_wheel_position)
        direction = math.copysign(1,slices)
        try:
            # Trying to catch connection exceptions here doesn't work. When attempting to reconnect, I get the error
            # failed turning wheel [Errno 2] No such file or directory
            # I confirmed that the /tmp socket file is missing that its trying to look for. Seems like a SDK bug.
            # This only and always happens when the "TimeoutError: Deadline exceeded. Connection lost" occurs below
            for i in range(abs(slices)):
                await wheel_motor.set_power(1/6*direction)
                current_wheel_position -= int(direction)
        except Exception as e:
            # Return the current position as the robot probably turned the last position before exception occurred.
            print("exception happened", e)
            return current_wheel_position, e
    return current_wheel_position, None

async def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--location-secret", required=True, type=str)
    parser.add_argument("--robot-address", required=True, type=str)
    args = parser.parse_args()
    print(args)

    # try:
    print("connecting to robot")
    robot = await connect(args.location_secret, args.robot_address)
    print("turning wheel to initial position 0")
    while True:
        try:
            if robot is None:
                robot = await connect(args.location_secret, args.robot_address)
            wheel_motor = Motor.from_robot(robot, "wheel_motor")
            for i in range(6):
                await wheel_motor.set_power(1/6)
            break
        except Exception as e:
            print("Failed to connect", e)
            await robot.close()
            robot = None
            continue

    current_wheel_position = 0
    while True:
        try:
            if robot is None:
                robot = await connect(args.location_secret, args.robot_address)
            wheel_motor = Motor.from_robot(robot, "wheel_motor")
            # This sometimes fails with the exception "TimeoutError: Deadline exceeded. Connection lost".
            # Seems like the connection is lost while controlling the robot. Even with short gRPC calls
            # Concerning: The robot continues to operate to the desired position but returns an exception early
            # To account for this I assume the robot did turn correctly and still set the wheel position.
            current_wheel_position, ex = await control_wheel(wheel_motor, current_wheel_position)
            print("wheel now at: ",current_wheel_position)
            if ex is not None:
                print("exception happened during turning, trying to recover. Exception: ", ex)
            else:
                time.sleep(15)
        except Exception as e:
            print("Failed to connect", e)
            await robot.close()
            robot = None
            continue
    # If I kill the script (ctrl-c) or an exception happens, I have difficultly connecting again. I have to reset the ESP32 every time.
    # If I don't reset the board, the script will attempt to connect 50 times with none successful.

if __name__ == '__main__':
    asyncio.run(main())

File: reader.go
package main

import (
    "context"
    "encoding/json"
    "fmt"
    "io/ioutil"
    "log"
    "math"
    "os"
    "time"

    "github.com/spf13/pflag"
    "golang.org/x/oauth2"
    "golang.org/x/oauth2/google"
    "google.golang.org/api/calendar/v3"
    "google.golang.org/api/option"
    viam "go.viam.com/rdk/components/motor"
	"go.viam.com/rdk/robot/client"
    "go.viam.com/utils/rpc"
	"go.viam.com/rdk/logging"
	"go.viam.com/rdk/components/motor"
)

const (
    SCOPES            = "https://www.googleapis.com/auth/calendar.readonly"
    TOKEN_FILE        = "token.json"
    CREDENTIALS_FILE  = "credentials.json"
    OUT_OF_OFFICE     = 0
    WORK_FROM_HOME    = 1
    GOING_TO_EVENT    = 2
    FOCUS_TIME        = 3
    AVAILABLE         = 4
    IN_MEETING        = 5
)

var eventTypeToWheelPosition = map[string]int{
    "outOfOffice": OUT_OF_OFFICE,
    "focusTime":   FOCUS_TIME,
    "default":     IN_MEETING,
}

func getCreds() (*oauth2.Config, *oauth2.Token, error) {
    b, err := ioutil.ReadFile(CREDENTIALS_FILE)
    if err != nil {
        return nil, nil, fmt.Errorf("unable to read client secret file: %v", err)
    }

    config, err := google.ConfigFromJSON(b, SCOPES)
    if err != nil {
        return nil, nil, fmt.Errorf("unable to parse client secret file to config: %v", err)
    }

    tokFile := TOKEN_FILE
    tok, err := tokenFromFile(tokFile)
    if err != nil {
        tok = getTokenFromWeb(config)
        saveToken(tokFile, tok)
    }

    return config, tok, nil
}

func tokenFromFile(file string) (*oauth2.Token, error) {
    f, err := os.Open(file)
    if err != nil {
        return nil, err
    }
    defer f.Close()
    tok := &oauth2.Token{}
    err = json.NewDecoder(f).Decode(tok)
    return tok, err
}

func getTokenFromWeb(config *oauth2.Config) *oauth2.Token {
    authURL := config.AuthCodeURL("state-token", oauth2.AccessTypeOffline)
    fmt.Printf("Go to the following link in your browser then type the authorization code: \n%v\n", authURL)

    var authCode string
    if _, err := fmt.Scan(&authCode); err != nil {
        log.Fatalf("Unable to read authorization code: %v", err)
    }

    tok, err := config.Exchange(context.TODO(), authCode)
    if err != nil {
        log.Fatalf("Unable to retrieve token from web: %v", err)
    }
    return tok
}

func saveToken(path string, token *oauth2.Token) {
    fmt.Printf("Saving credential file to: %s\n", path)
    f, err := os.Create(path)
    if err != nil {
        log.Fatalf("Unable to cache oauth token: %v", err)
    }
    defer f.Close()
    json.NewEncoder(f).Encode(token)
}

func getNextWheelPosition() (int, error) {
    config, tok, err := getCreds()
    if err != nil {
        return AVAILABLE, err
    }

    client := config.Client(context.Background(), tok)
    srv, err := calendar.NewService(context.Background(), option.WithHTTPClient(client))
    if err != nil {
        return AVAILABLE, fmt.Errorf("unable to retrieve Calendar client: %v", err)
    }

    now := time.Now().Format(time.RFC3339)
    events, err := srv.Events.List("primary").ShowDeleted(false).
        SingleEvents(true).TimeMin(now).MaxResults(1).OrderBy("startTime").
        Do()
    if err != nil {
        return AVAILABLE, fmt.Errorf("unable to retrieve next events: %v", err)
    }

    if len(events.Items) == 0 {
        fmt.Println("No upcoming events found.")
        return AVAILABLE, nil
    }

    event := events.Items[0]
    eventType := event.EventType
    fmt.Printf("Next event type: %s\n", eventType)
    start := event.Start.DateTime
    if start == "" {
        start = event.Start.Date
    }

    startTime, _ := time.Parse(time.RFC3339, start)
    if eventType == "workingLocation" {
        if event.Summary != "Office" {
            return WORK_FROM_HOME, nil
        }
    } else if startTime.After(time.Now().Add(5 * time.Minute)) {
        fmt.Println("Next event is > 5 min from now, so AVAILABLE")
        return AVAILABLE, nil
    } else if startTime.After(time.Now()) {
        fmt.Println("Next event is <= 5 min from now, so GOING_TO_EVENT")
        return GOING_TO_EVENT, nil
    }

    if pos, ok := eventTypeToWheelPosition[eventType]; ok {
        return pos, nil
    }

    return AVAILABLE, nil
}

func connect(apiKeyID, apiKey, robotAddress string) (*client.RobotClient, error) {
	logger := logging.NewDebugLogger("client")
	machine, err := client.New(
		context.Background(),
		robotAddress,
		logger,
		client.WithDialOptions(rpc.WithEntityCredentials( 
			apiKeyID,
			rpc.Credentials{
				Type:    rpc.CredentialsTypeAPIKey, 
				Payload: apiKey,
			})),
	)
	if err != nil {
		logger.Fatal(err)
		return nil, err
	}
	
	return machine, nil
}

func controlWheel(wheelMotor viam.Motor, currentWheelPosition int) (int, error) {
    nextWheelPosition, err := getNextWheelPosition()
    if err != nil {
        return currentWheelPosition, err
    }

    if currentWheelPosition != nextWheelPosition {
        fmt.Printf("Turning wheel from %d to position %d\n", currentWheelPosition, nextWheelPosition)
        slices := currentWheelPosition - nextWheelPosition
        direction := int(math.Copysign(1, float64(slices)))
        for i := 0; i < int(math.Abs(float64(slices))); i++ {
            if err := wheelMotor.SetPower(context.Background(), -float64(direction)/6, nil); err != nil {
                fmt.Println("Exception happened", err)
                return currentWheelPosition, err
            }
            currentWheelPosition -= direction
        }
    }

    return currentWheelPosition, nil
}

func main() {
    apiKeyID := pflag.String("api-key-id", "", "The key id of the api key")
	apiKey := pflag.String("api-key", "", "The api key")
    robotAddress := pflag.String("robot-address", "", "Address of the robot")
    pflag.Parse()

    if *apiKeyID == "" || *apiKey == "" || *robotAddress == "" {
        log.Fatal("api-key-id, api-key, and robot-address are required flags")
    }

    fmt.Println("Connecting to robot")
    robot, err := connect(*apiKeyID, *apiKey, *robotAddress)
    if err != nil {
        log.Fatalf("Failed to connect to robot: %v", err)
    }

    fmt.Println("Turning wheel to initial position 0")
	wheelMotor, err := motor.FromRobot(robot, "wheel_motor")
	if err != nil {
		log.Printf("Failed to find motor: %v", err)
		return
	}

	for i := 0; i < 6; i++ {
		_ = wheelMotor.SetPower(context.Background(), -1.0/6, nil)
	}

    currentWheelPosition := 0
    for {
        currentWheelPosition, err = controlWheel(wheelMotor, currentWheelPosition)
        if err != nil {
            log.Printf("Exception happened during turning, trying to recover: %v", err)
        } else {
            time.Sleep(1 * time.Minute)
        }
    }
}
